File: lastochka-desktop/src/lib/phone-utils.ts
/**
 * Утилиты для работы с номерами телефонов
 */

// Форматирование номера телефона для отображения
export function formatPhoneNumber(value: string): string {
  // Удаляем всё кроме цифр
  const digits = value.replace(/\D/g, '')
  
  // Если начинается с 8, заменяем на +7
  let normalized = digits
  if (normalized.startsWith('8')) {
    normalized = '7' + normalized.slice(1)
  }
  
  // Если не начинается с 7, добавляем 7
  if (!normalized.startsWith('7') && normalized.length > 0) {
    normalized = '7' + normalized
  }
  
  // Форматируем: +7 (XXX) XXX-XX-XX
  if (normalized.length <= 1) {
    return normalized ? `+${normalized}` : ''
  }
  if (normalized.length <= 4) {
    return `+${normalized.slice(0, 1)} (${normalized.slice(1)})`
  }
  if (normalized.length <= 7) {
    return `+${normalized.slice(0, 1)} (${normalized.slice(1, 4)}) ${normalized.slice(4)}`
  }
  if (normalized.length <= 9) {
    return `+${normalized.slice(0, 1)} (${normalized.slice(1, 4)}) ${normalized.slice(4, 7)}-${normalized.slice(7)}`
  }
  return `+${normalized.slice(0, 1)} (${normalized.slice(1, 4)}) ${normalized.slice(4, 7)}-${normalized.slice(7, 9)}-${normalized.slice(9, 11)}`
}

// Очистка номера для отправки на сервер
export function cleanPhoneNumber(value: string): string {
  const digits = value.replace(/\D/g, '')
  
  // Если начинается с 8, заменяем на 7
  let cleaned = digits
  if (cleaned.startsWith('8')) {
    cleaned = '7' + cleaned.slice(1)
  }
  
  // Если не начинается с 7, добавляем 7 (для России)
  if (!cleaned.startsWith('7') && cleaned.length > 0) {
    cleaned = '7' + cleaned
  }
  
  return cleaned
}

// Валидация номера телефона (российские номера)
export function isValidPhoneNumber(value: string): boolean {
  const cleaned = cleanPhoneNumber(value)
  // Российские номера: 7 + 10 цифр = 11 цифр
  return cleaned.length === 11 && /^7\d{10}$/.test(cleaned)
}

// Проверка, является ли номер российским
export function isRussianPhone(value: string): boolean {
  const cleaned = cleanPhoneNumber(value)
  return cleaned.startsWith('7') && cleaned.length === 11
}

// Получение маски для ввода
export function getPhoneMask(): string {
  return '+7 (999) 999-99-99'
}

/**
 * Утилиты для работы с email
 */

// Валидация email
export function isValidEmail(email: string): boolean {
  const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/
  return emailRegex.test(email.trim())
}

// Нормализация email (приведение к нижнему регистру, trim)
export function normalizeEmail(email: string): string {
  return email.trim().toLowerCase()
}

// Проверка домена email
export function isValidEmailDomain(email: string): boolean {
  const domain = email.split('@')[1]
  if (!domain) return false
  
  // Минимум 2 части (domain.tld)
  const parts = domain.split('.')
  return parts.length >= 2 && parts.every(p => p.length > 0)
}

File: lastochka-android/app/src/main/java/ru/lastochka/messenger/util/PhoneUtils.kt
package ru.lastochka.messenger.util

/**
 * Форматирование телефона в формат +7 (XXX) XXX-XX-XX
 * Возвращает Pair(formattedText, newCursorPos) для корректной работы курсора
 */
fun formatPhoneNumberWithCursor(text: String, cursorPosition: Int): Pair<String, Int> {
    // Извлекаем только цифры из всего текста
    val fullDigits = text.filter { it.isDigit() }

    // Если начинается с 8, заменяем на 7
    val cleaned = if (fullDigits.startsWith("8") && fullDigits.length >= 1) {
        "7" + fullDigits.substring(1)
    } else {
        fullDigits
    }

    // Ограничиваем 11 цифрами (7 + 10)
    val digits = if (cleaned.length > 11) cleaned.substring(0, 11) else cleaned

    // Извлекаем только цифры до позиции курсора в исходном тексте
    val beforeCursor = text.substring(0, cursorPosition.coerceAtMost(text.length)).filter { it.isDigit() }
    val digitsBeforeCursor = beforeCursor.length.coerceAtMost(digits.length)

    // Строим форматированную строку +7 (XXX) XXX-XX-XX
    val formatted = buildString {
        if (digits.isEmpty()) return@buildString

        append("+7")
        if (digits.length > 1) {
            append(" (")
            val end = 1 + (digits.length - 1).coerceAtMost(3)
            append(digits.substring(1, end))
        }
        if (digits.length > 4) {
            append(") ")
            val end = 4 + (digits.length - 4).coerceAtMost(3)
            append(digits.substring(4, end))
        }
        if (digits.length > 7) {
            append("-")
            val end = 7 + (digits.length - 7).coerceAtMost(2)
            append(digits.substring(7, end))
        }
        if (digits.length > 9) {
            append("-")
            val end = 9 + (digits.length - 9).coerceAtMost(2)
            append(digits.substring(9, end))
        }
    }

    // Вычисляем позицию курсора в отформатированной строке
    val newCursorPos = when {
        digitsBeforeCursor == 0 -> 0
        digitsBeforeCursor == 1 -> 2  // после "+7"
        digitsBeforeCursor <= 4 -> 4 + (digitsBeforeCursor - 1)  // внутри "(XXX"
        digitsBeforeCursor <= 7 -> 9 + (digitsBeforeCursor - 4)  // внутри " XXX"
        digitsBeforeCursor <= 9 -> 13 + (digitsBeforeCursor - 7) // внутри "-XX"
        else -> 16 + (digitsBeforeCursor - 9)                    // внутри "-XX"
    }

    // Не даём курсору выйти за пределы строки
    val clampedCursor = newCursorPos.coerceAtMost(formatted.length)

    return Pair(formatted, clampedCursor)
}

/**
 * Простое форматирование (без управления курсором)
 */
fun formatPhoneNumber(input: String): String {
    return formatPhoneNumberWithCursor(input, input.length).first
}

/**
 * Проверка валидности email
 */
fun isValidEmail(email: String): Boolean {
    if (email.isBlank()) return false
    val pattern = Regex("^[A-Za-z0-9+_.-]+@[A-Za-z0-9.-]+\\.[A-Za-z]{2,}$")
    return pattern.matches(email.trim())
}

/**
 * Проверка валидности телефона (11 цифр, начинается с 7)
 */
fun isValidPhoneNumber(phone: String): Boolean {
    val cleaned = phone.filter { it.isDigit() }
    return cleaned.length == 11 && cleaned.startsWith("7")
}

/**
 * Очистка номера телефона (только цифры)
 */
fun cleanPhoneNumber(phone: String): String {
    return phone.filter { it.isDigit() }
}
